### PR TITLE
Support threads and automatic history sync

### PR DESCRIPTION
**Pull request by nicois**
Threading information is captured, but only basic formatting is done, at
the moment. More thought is needed for the best way of presenting
matches in threaded conversations.

Also improved the gitignore to not include venv files, and removed CRLF
endings in the source files.

### CODE REVIEW

**Comment by nicois** [comment]
Something I'm thinking about, which would be affected by threads, is showing context around the matched entries. ie: showing the X entries before/after the machines lines, or X minutes. Similar to diff context lines. But you want to group the answers by thread, and present it in a way which makes sense to the viewer

**Comment by nicois** [comment]
In addition to some miscellaneous improvements, this now will use channel.history to automatically fetch gaps in the history. This will work when a channel is added, to pull in any history, but also if the server goes offline. The RTC interface will lose any messages sent while offline, but channel.history will fill the gap.

**Comment by alex-shpak** [comment]
Hi!
Looks great! I will test it a bit and come back to you.

File: archivebot.py [hunk 1]
```diff
@@ -154,79 +178,153 @@ def handle_query(event):
...
+    else:
+        return '*<@%s> <#%s> <!date^%s^{date_short} {time_secs}|date>*\n%s)' % (username, line['channel'], timestamp, message)
+
+
+def update_channel_history(conn):
+    """
+    For each channel we have previously received, check if there are any later messages
+    which we missed
+    """
+    channels_map = dict(record for record in conn.execute("SELECT channel, MAX(timestamp) as latest_timestamp FROM messages"))
```
Comment (alex-shpak) [inline archivebot.py]: Shouldn't it be `DISTINCT(channel)`?

**Comment by nicois** [comment]
no, but there should have been a GROUP BY.

please review my latest push.

Thanks

**Pull request merged**